src/README.md

```diff
@@ -0,0 +1,26 @@
+This README serves as developer documentation for CSV.jl internals. It doesn't pretend to be comprehensive, but was created with the aim to explain both the overall strategies CSV.jl uses in parsing delimited files as well as the trickiest mechanics and some of the reasoning that went into the architecture.
+
+# CSV.File
+
+Let's go over details specific to `CSV.File`.
+
+## The `pool` keyword argument
+
+CSV.jl provides a native integration with the [PooledArrays.jl](https://github.com/JuliaData/PooledArrays.jl/) package, which provides an array storage optimization by having a (hopefully) small pool of "exspensive" (big or heap-allocated, or whatever) values, along with a memory-efficient integer array of "refs" where each ref maps to one of the values in the pool. This is sometimes referred to as a "dictionary encoding" in various data formats. As an example, if you have a column with 1,000,000 elements, but only 10 unique string values, you can have a `Vector{String}` pool to store the 10 unique strings and give each a unique `UInt32` value, and a `Vector{UInt32}` "ref array" for the million elements, where each element just indexes into the pool to get the actual string value.
+
+By providing the `pool` keyword argument, users can control how this optimization will be applied to individual columns, or to all columns of the delimted text being read.
+
+Valid inputs for `pool` include:
+  * A `Bool`, `true` or `false`, which will apply to _all_ columns parsed; they either will _all_ be pooled, or _all_ not pooled
+  * A `Real`, which will be converted to `Float64`, which should be a value between `0.0` and `1.0`, to indicate the % cardinality threshold _under which_ a column will be pooled. e.g. by passing `pool=0.1`, if a column has less than 10% unique values, it will end up as a `PooledArray`, otherwise a normal array. Like the `Bool` argument, this will apply the same % threshold to _all_ columns
+  * As a special-case, `NaN` (the `Float64` not-a-number value), is used to signal "default" pooling behavior. "Default" here means that only columns with `String` values will be considered for pooling, and then a different threshold will be considered based on whether a file is being parsed by multiple threads or not; for single-threaded parsing, the default threshold for pooling is `0.75`. For multi-threaded parsing, the default threshold is `0.25`. The reasoning here is that single-threaded parsing scenarios are almost always smaller files, where the relative cost of pooling is smaller to the benefits of a pooled column output. For multithreaded parsing, pooled reference values must be synchronized across threads, and it can be costly to "unpool" a column after parsing the entire file.
+  * An `AbstractVector`, where the # of elements should/needs to match the # of columns in the dataset. Each element of the `pool` argument should be a `Bool` or `Real` indicating the pooling behavior for each specific column.
```

Comment:
why "should/needs" - I understand it is "must" (if it does not match we throw an error - right?)

Reply:
Well, this was a question I had while implementing the "column widening" feature. If a user provides `header` or `types`, or other keyword argument for each column, should we accept no matter what? And then when parsing data, emit warnings on too few columns if necessary? The use-case I have in mind is that if you have a header row w/ too few column names, but you know actual data has _more_ columns, you might provide `header`/`types`/`pool` keyword arguments for all the columns _in the data_, and then we ignore the header row in the data. So this would mean we accept any length `header`/`types` `AbstractVector` keyword arguments, trusting the user knows how many columns in the data. I'm not totally sure this is the right thing to do though, since user may get it wrong.

Reply:
OK - since this "should/needs" is because of just in case handling of column widening it is OK. What do you do now when you detect column widening? Emit a warning?

Reply:
We don't emit a warning, but we probably should. Previous to this PR, we emit a warnign when too many columns are detected for row, now we parse them. I think it would probably be good to emit a warning though since it's rare behavior and probably unexpected.

Reply:
I'm pretty against throwing errors once we've started parsing, except for the one fatal case where a cell is not quoted properly. In my experience, and from what I've heard from others, it's really annoying to have parsing fail, especially in the case where it's a huge file, so it's parsing for a long time, then gets to the end and there happens to be a junk row that causes the whole parsing to fail. I'd be fine though with adding a "don't widen" keyword arg if we find out people actually _don't_ want this extra data parsed.

Reply:
I can understand this, therefore if you prefer this I would vote for three valued logic as proposed above:
* `nothing`: warning (the default)
* `true` or `false` - accept or throw an error

All three behaviors might be expected by users in different use cases.